### DISCUSSION TITLE
Does Clickhouse read unnecessary data?

### DISCUSSION THREAD

**Discussion by SergioWhite** [ClickHouse/ClickHouse#72389] [Q&A]
Hello! 

I am using ClickHouse version 23.1.7.30.

Let's create a simple table and fill it with data:
```
create table my_table (id UUID, ts timestamp, data String) engine = MergeTree order by ts;
insert into my_table select generateUUIDv4(), now() - rand() % 31536000, randomString(100) from numbers(10000000);
```
And let's try to execute two queries:
```
1. select * from my_table where ts between now() - 30000000 and now() limit 500
2. select * from my_table where (ts, id) in (select ts, id from my_table where ts between now() - 30000000 and now() limit 500)
```
As we can see, the queries differ in that in the second case we do not use * to get subdata. If we look at the query logs in the query_log table, we will see something like this:

![image](https://github.com/user-attachments/assets/fe5d3760-3c51-46d4-b6cd-361d4c24a7d9)

We can see that the query with the subquery consumes less memory (22 MB versus 70 MB). Is this true, am I interpreting the results correctly? If so, please explain why the first query consumes more memory. And is there any way to optimize the first query so that it produces similar memory consumption figures as the second query?

**Comment by den-crane** [comment]
Yes, Clickhouse executes the first query sub-optimaly.

* https://github.com/ClickHouse/ClickHouse/issues/45868

* https://github.com/ClickHouse/ClickHouse/pull/55518 (not merges yet)
